### PR TITLE
operation Contract

### PR DESCRIPTION
**Pull request by NikoBergemann**
- approvalContract --> operationContract
- adjust operationContract to new Requirements
- provide public scala Interface for Operations

### CODE REVIEW

File: hlf/scala/operation.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Hyperledger Scala API Operation
+
+In the following, the hlf-api for the "operation" handling will be explained.
+For insight in the General hlf-api, please refer to it's .Readme - File.
+> [!NOTE]
+Any Method can throw these [Exceptions](errors.md#Exceptions).
+
+For "admission" handling we offer the following interface.
```
Comment (anemonekk) [inline hlf/scala/operation.md]: change to "operation"

Comment (NikoBergemann) [reply hlf/scala/operation.md]:
```suggestion
For "operation" handling we offer the following interface. 

```

File: hlf/chaincode/operation.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    - [GenericError](errors.md#GenericError) 
+      ```json
+      {
+        "type": "HLUnprocessableLedgerState",
+        "title": "The state on the ledger does not conform to the specified format"
+      }
+      ```
+      - Description: This error is returned, if the state of data on the ledger is not consistent with the current model. This error should only occur if the model changes while the old ledger state remains without modification.
+
```
Comment (NikoBergemann) [inline hlf/chaincode/operation.md]:
```suggestion
### GetOperations
- ID = getOperations
- Send
    - enrollmentId :: String (optional)
    - state :: String (optional)
- Receive
    - operationsList :: List\<[OperationData](#OperationData)\>
      - Description: Returns the full list of existing Operations matching the filter parameters.
        If these parameters are empty, the exhaustive list of all existing Operations is returned.

```
Remove faulty error. If a single operation could not be rea we still want to return as many "working" operationData as possible

File: hlf/scala/operation.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+- Returns
+    - OperationsList :: List\<String Json (refer to: [OperationData](../chaincode/operation.md#OperationData))\> 
+        - => Exhaustive List of all Operations, filtered by
+            inputs "enrollmentId", "state".
+            Empty if no match could be found.
+
+> Note! Invalid States on the ledger get ignored.
+> We return all valid Operations that match the filters.
+
+## GetOperation(operationId: String)
```
Comment (anemonekk) [inline hlf/scala/operation.md]:
```suggestion
## GetOperationData(operationId: String)

```

**Review by bastihav — APPROVED** [review]

File: hlf/chaincode/operation.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+In the following, the chaincode api for the "Operation"-Chaincode is described.
+This contains all Transactions and Models provided for this domain.
+The Errors returned are defined [here](errors.md#Errors).
+
+## Transactions
+
+### ApproveTransaction
+- ID = approveTransaction
+- Send
+    - contractName, transactionName, param1, ..., paramN :: String
```
Comment (matthias-geuchen) [inline hlf/chaincode/operation.md]:
```suggestion
    - contractName, transactionName :: String
    - params :: List\<String\>

```

File: hlf/chaincode/operation.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    - [GenericError](errors.md#GenericError) 
+      ```json
+      {
+        "type": "HLUnprocessableLedgerState",
+        "title": "The state on the ledger does not conform to the specified format"
+      }
+      ```
+      - Description: This error is returned, if the state of data on the ledger is not consistent with the current model. This error should only occur if the model changes while the old ledger state remains without modification.
+
```
Comment (matthias-geuchen) [inline hlf/chaincode/operation.md:53]:
```suggestion

- [GenericError](errors.md#GenericError) 
      ```json
      {
        "type": "HLNotFound",
        "title": "There is no Operation for the given operationId"
      }
      ```
      - Description: This error is returned, if there is no operation for the specified operationId present on the ledger.
```

**Review by mco33415 — APPROVED** [review]

File: hlf/scala/operation.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+## RejectTransaction(operationId: String, rejectMessage: String)
+- Returns
+    - OperationData :: Json (refer to: [OperationData](../chaincode/operation.md#OperationData))
+        - => Success
+    - TransactionError :: Json (refer to: [DetailedError](../chaincode/errors.md#DetailedError))
+        - => error is returned
+          - If the given parameters could not be parsed.
+    - TransactionError :: Json (refer to: [GenericError](../chaincode/errors.md#GenericError))
+        - => error is returned
```
Comment (matthias-geuchen) [inline hlf/scala/operation.md:29]:
```suggestion
        - => error is returned
          - If the operationId is not present on the ledger.
```

File: hlf/scala/operation.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            - This error should only occurr if the model changes while the old ledger state remains without modification.
+
+## GetOperations(enrollmentId: String, state: String)
+Gets the full List of existing Operations.
+Applies filters to match enrollmendId, state.
+If any of theses parameter is empty, its filter will not be applied.
+- Returns
+    - OperationsList :: List\<String Json (refer to: [OperationData](../chaincode/operation.md#OperationData))\> 
+        - => Exhaustive List of all Operations, filtered by
+            inputs "enrollmentId", "state".
```
Comment (matthias-geuchen) [inline hlf/scala/operation.md]:
```suggestion
            inputs "operationId", "state".

```

File: hlf/scala/operation.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        - => Success
+    - TransactionError :: Json (refer to: [DetailedError](../chaincode/errors.md#DetailedError))
+        - => error is returned
+          - If the given parameters could not be parsed.
+    - TransactionError :: Json (refer to: [GenericError](../chaincode/errors.md#GenericError))
+        - => error is returned
+          - If the state of data on the ledger is not consistent with the curent model.
+            - This error should only occurr if the model changes while the old ledger state remains without modification.
+
+## GetOperations(enrollmentId: String, state: String)
```
Comment (matthias-geuchen) [inline hlf/scala/operation.md]:
```suggestion
## GetOperations(operationId: String, state: String)

```

File: hlf/chaincode/operation.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+       For detailed informations see [Input Checks](#Checks).
+
+
+### GetOperationData
+- ID = getOperationData
+- Send
+    - operationId :: String
+- Receive
+    - operationData :: [OperationData](#OperationData)
+
```
Comment (matthias-geuchen) [inline hlf/chaincode/operation.md:85]:
```suggestion

- [GenericError](errors.md#GenericError) 
      ```json
      {
        "type": "HLNotFound",
        "title": "There is no Operation for the given operationId"
      }
      ```
      - Description: This error is returned, if there is no operation for the specified operationId present on the ledger.
```

File: hlf/scala/operation.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+## GetOperationData(operationId: String)
+
+- Returns
+    - OperationData :: Json (refer to: [OperationData](../chaincode/operation.md#OperationData))
+        - => Success
+    - TransactionError :: Json (refer to: [DetailedError](../chaincode/errors.md#DetailedError))
+        - => error is returned
+          - If the given parameters could not be parsed.
+    - TransactionError :: Json (refer to: [GenericError](../chaincode/errors.md#GenericError))
+        - => error is returned
```
Comment (matthias-geuchen) [inline hlf/scala/operation.md:58]:
```suggestion
        - => error is returned
          - If the operationId is not present on the ledger.
```

**Review by matthias-geuchen — CHANGES_REQUESTED** [review]

File: hlf/chaincode/operation.md [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+{
+  "operationId" : "0424974c68530290458c8d58674e2637f65abc127057957d7b3acbd24c208f93",
+  "transactionInfo" : {
+    "contractName" : "UC4.Certificate",
+    "transactionName" : "addCertificate",
+    "parameters": [
+      "EnrollmentID_001",
+      "legit_certificate"
+    ]
+  },
```
Comment (NikoBergemann) [inline hlf/chaincode/operation.md]:
```suggestion
    "parameters": "[
      \"EnrollmentID_001\",
      \"legit_certificate\"
    ]"
  },

```

File: hlf/chaincode/operation.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+       For detailed informations see [Input Checks](#Checks).
+
+### RejectTransaction
+- ID = rejectTransaction
+- Send
+    - operationId :: String
+    - rejectMessage :: String
+- Receive
+    - operationData :: [OperationData](#OperationData) 
+      -  Description: Success, returns the list of approvals.
```
Comment (anemonekk) [inline hlf/chaincode/operation.md]:
```suggestion
      -  Description: Success, returns the list of operations.

```

**Review by RBeierling — APPROVED** [review]

File: hlf/chaincode/operation.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+
+### <a id="TransactionInfo" />TransactionInfo
+```json
+{
+  "contractName" : "UC4.Certificate",
+  "transactionName" : "addCertificate",
+  "parameters": [
+    "EnrollmentID_001",
+    "legit_certificate"
+  ]
```
Comment (matthias-geuchen) [inline hlf/chaincode/operation.md]:
```suggestion
  "parameters": "[
    \"EnrollmentID_001\",
    "\legit_certificate\"
  ]"

```

**Review by matthias-geuchen — APPROVED** [review]
Looks good, apart from a small inconsistency between OperationData and TransactionInfo models.

**Pull request merged**